aggregator/leader_flush_mgr.go

```diff
@@ -215,26 +222,41 @@ func (mgr *leaderFlushManager) computeNextFlushNanos(flushInterval time.Duration
 	return nextFlushNanos
 }
 
+func (mgr *leaderFlushManager) collectFlushTimesWithLock(
+	buckets []*flushBucket,
+) {
+	for _, bucket := range buckets {
+		for _, flusher := range bucket.flushers {
+			flusherID := flusherID{
+				shard:      flusher.Shard(),
+				resolution: flusher.Resolution(),
+			}
+			mgr.flushedNanos[flusherID] = flusher.LastFlushedNanos()
+		}
+	}
+}
+
 func (mgr *leaderFlushManager) prepareFlushTimesWithLock(
 	buckets []*flushBucket,
 ) schema.ShardSetFlushTimes {
+	// Collect the flush times from all the flushers in the buckets.
+	mgr.collectFlushTimesWithLock(buckets)
```

Comment:
nit: it seems a little odd to me that `flushedNanos` is a field on the `leaderFlushManager` even though it is only used once, on line 249. Perhaps it would clearer for `collectFlushTimesWithLock ` to return the map explicitly?

Reply:
The `flushedNanos` field is to retain all the flushed times that the leader has ever seen, including the flush times of shards that have been closed. This is so that during a topology change, if the leader closes out a shard, the flush times of that shard will still be persisted in kv for the follower to receive so the follower can flush and close the shard accordingly. 

Otherwise, there is a race condition during a topology change where the leader closes a leaving shard and flushes, and the flush times persisted in kv will no longer include the latest flush times of the closed shard, which means the follower won't ever close that leaving shard due to missing flush times of that shard.

Reply:
Oh, I see, so `Prepare` and therefore `collectFlushTimesWithLock ` will be called repeatedly?

Reply:
Correct, `Prepare` is called every second by default by the flush manager (https://github.com/m3db/m3aggregator/blob/master/aggregator/flush_mgr.go#L282)